### PR TITLE
HIVE-26121: Hive transaction rollback should be thread-safe

### PR DESCRIPTION
**Pull request by deniskuzZ**
### Why are the changes needed?

Proper ACID rollback handling in case of operation interruption

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Itest

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java [hunk 1]
```diff
@@ -570,7 +570,7 @@ void endTransactionAndCleanup(boolean commit) throws LockException {
     txnRollbackRunner = null;
   }
 
-  void endTransactionAndCleanup(boolean commit, HiveTxnManager txnManager) throws LockException {
+  synchronized void endTransactionAndCleanup(boolean commit, HiveTxnManager txnManager) throws LockException {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java:573]: Could we leave a comment here, why is this synchronized?

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 2]
```diff
@@ -710,49 +691,32 @@ private Heartbeater startHeartbeat(long initialDelay) throws LockException {
...
-          if (now - startTime > 30000) {
-            LOG.warn("Heartbeat task cannot be cancelled for unknown reason. QueryId: " + queryId);
-            break;
-          }
-          try {
-            Thread.sleep(sleepInterval);
-          } catch (InterruptedException e) {
-          }
-          sleepInterval *= 2;
+  private synchronized void stopHeartbeat() {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:694]: Comment here as well

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:697]: added

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 3]
```diff
@@ -574,30 +571,24 @@ public void rollbackTxn() throws LockException {
...
-
-      // Re-checking as txn could have been closed, in the meantime, by a competing thread.
-      if (isTxnOpen()) {
-        if (replPolicy != null) {
-          getMS().replRollbackTxn(txnId, replPolicy, TxnType.DEFAULT);
-        } else {
-          getMS().rollbackTxn(txnId);
-        }
+      
+      if (replPolicy != null) {
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:580]: marked as @NotThreadSafe

File: ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java:576]: added

**Comment by pvary** [comment]
I have missed this before, but do we really need to synchronize `DriverTxnHandler.endTransactionAndCleanup` and `DbTxnManager.java.stopHeartbeat` too?

Otherwise LGTM

**Comment by deniskuzZ** [comment]
> endTransactionAndCleanup

yes, we definitely need to synchronize `DriverTxnHandler.endTransactionAndCleanup`.  As for  `DbTxnManager.java.stopHeartbeat` it was already kinda synchronized and it's possible that concurrent execution might still occur when shutdownhook is invoked.

**Review by pvary — APPROVED** [review]

**Pull request merged**